usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2939,6 +2939,125 @@ static CK_RV ccatok_pkey_check(STDLL_TokData_t *tokdata, SESSION *session,
+    rc = object_unlock(key_obj);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("object_unlock failed, rc=0x%lx\n", rc);
+        goto done;
+    }
+
+    rc = object_lock(key_obj, WRITE_LOCK);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not obtain write lock.\n");
+        goto done;
+    }
+
+    /*
+     * The key_obj could be modified between unlock and lock. Therefore get
+     * the attribute when we have the write lock here.
+     */
+    if (template_attribute_get_non_empty(key_obj->template,
+                                         CKA_IBM_OPAQUE, &skey_attr) != CKR_OK) {
+        TRACE_ERROR("This key has no blob: should not occur!\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto unlock;
+    }
+
+retry:
+    /* Convert secure key to protected key. */
+    rc = ccatok_pkey_skey2pkey(tokdata, skey_attr, &pkey_attr, xts_mode);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("protkey creation failed, rc=0x%lx\n", rc);
+        goto unlock;
+    }
+
+    /*
+     * In case of XTS, check if the wkvp's of the two keys are identical.
+     * An LGR could have happened between the creation of the two keys.
+     */
+    if (xts_mode) {
+        vp_offset1 = pkey_attr->ulValueLen / 2 - PKEY_MK_VP_LENGTH;
+        vp_offset2 = pkey_attr->ulValueLen - PKEY_MK_VP_LENGTH;
+        if (memcmp((CK_BYTE *)pkey_attr->pValue + vp_offset1,
+                   (CK_BYTE *)pkey_attr->pValue + vp_offset2,
+                   PKEY_MK_VP_LENGTH) != 0) {
+            num_retries++;
+            if (num_retries < PKEY_CONVERT_KEY_RETRIES) {
+                TRACE_DEVEL("%s vp of xts key 1 does not match with vp of "
+                            "xts key 2, retry %d of %d ...\n",
+                            __func__, num_retries, PKEY_CONVERT_KEY_RETRIES);
+                goto retry;
+            }
+            rc = CKR_FUNCTION_FAILED;
+            goto unlock;
+        }
+    }
+
+    /*
+     * Save new protkey attr in key_obj. This happens only in memory,
+     * works also for r/o sessions.
+     */
+    rc = template_update_attribute(key_obj->template, pkey_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed, rc=0x%lx\n", rc);
+        goto unlock;
+    }
+
+    /* If we have a r/w session, also save obj to disk. */
+    if (ccatok_pkey_session_ok_for_obj(session, key_obj)) {
+        if (object_is_token_object(key_obj)) {
+            rc = object_mgr_save_token_object(tokdata, key_obj);
+            if (rc != CKR_OK) {
+                TRACE_ERROR("Could not save token obj to repository, rc=0x%lx.\n", rc);
+                goto unlock;
+            }
+        }
+    }
+
+    /* Update wkvp in CCA private data. */
+    memcpy(&cca_private->pkey_mk_vp,
+           (CK_BYTE *)pkey_attr->pValue + pkey_attr->ulValueLen - PKEY_MK_VP_LENGTH,
+           PKEY_MK_VP_LENGTH);
+
+    /* Pass back new protkey. Need to do this before unlocking the obj. */
+    memcpy(protkey, pkey_attr->pValue, pkey_attr->ulValueLen);
```

Comment:
Instead of checking for `*protkey_len < MAXPROTKEYSIZE - 16` at the beginning of the function, I would rather check here for  `*protkey_len < pkey_attr->ulValueLen` and return here with CKR_BUFFER_TOO_SMALL if it is too small.
While we know that `pkey_attr->ulValueLen` will most likely not be larger than `MAXPROTKEYSIZE - 16`, ist still better to check here.

Reply:
fixed.